python/activity_stream/activity_stream.py

```diff
@@ -0,0 +1,552 @@
+        Process new activity ids as they arrive from
+        the data manager.
+        
+        :param activity_ids: List of activity ids
+        """
+        self._bundle.log_debug("Process new data slot called "
+                            "for %s activity events" % len(activity_ids))
+                
+        # remove the "loading please wait .... widget
+        self._clear_loading_widget()
+        
+        
+        # note! For an item which hasn't been previously been cached or
+        # hasn't been visited for some time, there may be a lot more than
+        # MAX_STREAM_LENGTH updates. In this case, truncate the stream.
+        # this will result in a UI where you may have a maxmimum of 
+        # MAX_STREAM_LENGTH * 2 items (already loaded + new) and there
+        # may be gaps in activity data because we always want to show
+        # the latest data, so when we cull, it happens in the 'middle'
+        # of the stream, resulting in existing data, the a potential gap
+        # and then MAX_STREAM_LENGTH items.
+        #
+        # Note that this is in the UI only, so a refresh of the page 
+        # would immediately rectify the discrepancy.  
+        
+        # load in the new data
+        # the list of ids is delivered in ascending order
+        # and we pop them on to the widget
+
+        if len(activity_ids) > self.MAX_STREAM_LENGTH:
+            self._bundle.log_debug("Capping the %s new activity items down to "
+                                "%s items" % (len(activity_ids), self.MAX_STREAM_LENGTH))
+        
+            # transform [10,11,12,13,14,15,16,17] -> [15,16,17]
+            activity_ids = activity_ids[-self.MAX_STREAM_LENGTH:]
+        
+        for activity_id in activity_ids:
+            self._bundle.log_debug("Creating new widget...")
+            w = self._create_activity_widget(activity_id)
+            if w:            
+                self._widgets[activity_id] = w
+                self._bundle.log_debug("Adding %s to layout" % w)
+                self.ui.activity_stream_layout.addWidget(w)        
+                # add special blue border to indicate that this is a new arrival
+                w.setStyleSheet("QFrame#frame{ border: 1px solid rgba(48, 167, 227, 50%); }")
+        
+        # when everything is loaded in, load the thumbs
+        self._bundle.log_debug("Requesting thumbnails")
+        for activity_id in activity_ids:
+            self._data_manager.request_activity_thumbnails(activity_id)
+                
+        self._bundle.log_debug("Process new data complete.")
+
+        # turn off the overlay in case it is spinning
+        # (which only happens on a full load)
+        self.__overlay.hide()
+            
+    def _process_thumbnail(self, data):
+        """
+        New thumbnail has arrived from the data manager
+        """
+        # broadcast to all activity widgets
+        for widget in self._widgets.values():
+            widget.set_thumbnail(data)
```

Comment:
At first glance, this was confusing to me. Why would we set the same thumbnail for all widgets? But I think this is forwarding thumbnail data to all the widgets in case they need to display it? Not a huge thing, but maybe a verb other than `set`? Maybe `apply`?